cmd/agent/main.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/golang/glog"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/spf13/cobra"
+
+	proxyconfig "istio.io/api/proxy/v1/config"
+	"istio.io/pilot/cmd"
+	"istio.io/pilot/proxy"
+	"istio.io/pilot/proxy/envoy"
+	"istio.io/pilot/tools/version"
+)
+
+var (
+	meshconfig string
+	sidecar    proxy.Sidecar
+
+	rootCmd = &cobra.Command{
+		Use:   "agent",
+		Short: "Istio Pilot agent",
+		Long:  "Istio Pilot provides management plane functionality to the Istio service mesh and Istio Mixer.",
+	}
+
+	proxyCmd = &cobra.Command{
+		Use:   "proxy",
+		Short: "Envoy proxy agent",
+		RunE: func(c *cobra.Command, args []string) error {
+			// set values from environment variables
+			if sidecar.IPAddress == "" {
+				sidecar.IPAddress = os.Getenv("POD_IP")
+			}
+			if sidecar.InstanceName == "" {
+				sidecar.InstanceName = os.Getenv("POD_NAME")
+			}
+			if sidecar.InstanceNamespace == "" {
+				sidecar.InstanceNamespace = os.Getenv("POD_NAMESPACE")
+			}
+
+			glog.V(2).Infof("version %s", version.Line())
+
+			// receive mesh configuration
+			mesh, err := cmd.ReadMeshConfig(meshconfig)
+			if err != nil {
+				return multierror.Prefix(err, "failed to read mesh configuration.")
+			}
+
+			glog.V(2).Infof("mesh configuration %s", spew.Sdump(mesh))
+
+			var role proxy.Role = sidecar
+			if len(args) > 0 {
+				switch args[0] {
```

Comment:
3 processes would be cleaner, indeed. It's also not clear we'll even need this, but I suspect
at least egress might be better if it's local/close to the VMs. We can worry about it later, I didn't realize the configs are not 'additive'